syntax="proto2";

package aap_protobuf.shared;

/**
 * @file MessageStatus.proto
 * @brief Android Auto protocol status and error codes.
 *
 * MessageStatus enumerates all possible outcomes for Android Auto protocol
 * operations. Used in response messages to indicate success or specific failure
 * reason. Status codes are negative for errors, positive for success variants.
 *
 * Success statuses:
 *   - STATUS_SUCCESS: Operation completed successfully
 *   - STATUS_UNSOLICITED_MESSAGE: Message arrived without request (e.g., unsolicited nav update)
 *
 * Authentication failures (certificate validation, SSL negotiation):
 *   - STATUS_CERTIFICATE_ERROR: Certificate validation failed
 *   - STATUS_AUTHENTICATION_FAILURE: SSL/TLS handshake failed
 *   - STATUS_AUTHENTICATION_FAILURE_CERT_NOT_YET_VALID: Certificate validity period not started
 *   - STATUS_AUTHENTICATION_FAILURE_CERT_EXPIRED: Certificate has expired
 *
 * Channel/Service errors (invalid routing):
 *   - STATUS_INVALID_SERVICE: Requested service not available on device
 *   - STATUS_INVALID_CHANNEL: Channel ID (0-7) out of range
 *   - STATUS_INVALID_PRIORITY: Message priority level not supported
 *
 * Bluetooth errors (pairing, HFP handsfree profile):
 *   - STATUS_BLUETOOTH_UNAVAILABLE: Bluetooth radio disabled or not available
 *   - STATUS_BLUETOOTH_PAIRING_DELAYED: Device pairing in progress
 *   - STATUS_BLUETOOTH_HFP_ANOTHER_CONNECTION: HFP already connected to different device
 *   - STATUS_BLUETOOTH_HFP_CONNECTION_FAILURE: HFP connection attempt failed
 *   - STATUS_BLUETOOTH_INVALID_ADDRESS: Bluetooth MAC address malformed
 *
 * Resource errors (buffer overflows, timeouts):
 *   - STATUS_OUT_OF_MEMORY: Device memory exhausted
 *   - STATUS_BUSY: Device busy with other operations
 *   - STATUS_PING_TIMEOUT: Keep-alive ping response not received (connection stale)
 *
 * Scenario: Navigation connection failure handling:
 *   - Head unit sends OPEN_CHANNEL request for Navigation service
 *   - Device responds with MessageStatus CHANNEL_OPENED, status=INVALID_SERVICE
 *   - Head unit logs error, prompts user: "Navigation not available"
 *   - User cancels navigation, displays media instead
 */
enum MessageStatus {
  STATUS_UNSOLICITED_MESSAGE = 1;
  STATUS_SUCCESS = 0;
  STATUS_NO_COMPATIBLE_VERSION = -1;
  STATUS_CERTIFICATE_ERROR = -2;
  STATUS_AUTHENTICATION_FAILURE = -3;
  STATUS_INVALID_SERVICE = -4;
  STATUS_INVALID_CHANNEL = -5;
  STATUS_INVALID_PRIORITY = -6;
  STATUS_INTERNAL_ERROR = -7;
  STATUS_MEDIA_CONFIG_MISMATCH = -8;
  STATUS_INVALID_SENSOR = -9;
  STATUS_BLUETOOTH_PAIRING_DELAYED = -10;
  STATUS_BLUETOOTH_UNAVAILABLE = -11;
  STATUS_BLUETOOTH_INVALID_ADDRESS = -12;
  STATUS_BLUETOOTH_INVALID_PAIRING_METHOD = -13;
  STATUS_BLUETOOTH_INVALID_AUTH_DATA = -14;
  STATUS_BLUETOOTH_AUTH_DATA_MISMATCH = -15;
  STATUS_BLUETOOTH_HFP_ANOTHER_CONNECTION = -16;
  STATUS_BLUETOOTH_HFP_CONNECTION_FAILURE = -17;
  STATUS_KEYCODE_NOT_BOUND = -18;
  STATUS_RADIO_INVALID_STATION = -19;
  STATUS_INVALID_INPUT = -20;
  STATUS_RADIO_STATION_PRESETS_NOT_SUPPORTED = -21;
  STATUS_RADIO_COMM_ERROR = -22;
  STATUS_AUTHENTICATION_FAILURE_CERT_NOT_YET_VALID = -23;
  STATUS_AUTHENTICATION_FAILURE_CERT_EXPIRED = -24;
  STATUS_PING_TIMEOUT = -25;
  STATUS_COMMAND_NOT_SUPPORTED = -250;
  STATUS_FRAMING_ERROR = -251;
  STATUS_UNEXPECTED_MESSAGE = -253;
  STATUS_BUSY = -254;
  STATUS_OUT_OF_MEMORY = -255;
}